File: leetcode-04/solutions/python3/0090.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = []

        nums.sort()
        self.dfs(nums, len(nums), 0, [], ans)

        return ans

    def dfs(self, nums: List[int], target: int, s: int, path: List[int], ans: List[List[int]]) -> None:
        if target < 0:
            return

        ans.append(path)

        for i in range(s, len(nums)):
            if i > s and nums[i] == nums[i - 1]:
                continue
            self.dfs(nums, target - 1, i + 1, path + [nums[i]], ans)

File: leetcode-04/solutions/cpp/0090.cpp
class Solution {
 public:
  vector<vector<int>> subsetsWithDup(vector<int>& nums) {
    vector<vector<int>> ans;
    vector<int> path;

    sort(nums.begin(), nums.end());
    dfs(nums, nums.size(), 0, path, ans);

    return ans;
  }

 private:
  void dfs(vector<int>& nums, int target, int s, vector<int>& path,
           vector<vector<int>>& ans) {
    if (target < 0) return;

    ans.push_back(path);

    for (int i = s; i < nums.size(); i++) {
      if (i > s && nums[i] == nums[i - 1]) continue;
      path.push_back(nums[i]);
      dfs(nums, target - 1, i + 1, path, ans);
      path.pop_back();
    }
  }
};